File: pdf_parser.py
import pdfplumber
import re
from typing import Dict, List, Any, Optional, Tuple

class STUMPSPDFParser:
    """Parse STUMPS cricket scorecard PDFs using regex patterns"""
    
    def __init__(self, pdf_path: str):
        self.pdf_path = pdf_path
        self.pages = []
        self.match_info = {}
        self.batting_stats = []
        self.bowling_stats = []
        
    def extract_text(self) -> str:
        """Extract all text from PDF"""
        text = ""
        with pdfplumber.open(self.pdf_path) as pdf:
            for page in pdf.pages:
                text += page.extract_text() or ""
                text += "\n---PAGE_BREAK---\n"
        return text
    
    def parse_match_info(self, text: str) -> Dict[str, Any]:
        """Extract match information from page 1"""
        info = {}
        
        # Match title and result
        result_match = re.search(r'(\w+.*?)\s+(\d+)-(\d+)in([\d.]+)overs\s+(\w+.*?)\s+won\s+by\s+(\d+)\s+(wickets|runs)', text, re.IGNORECASE)
        if result_match:
            info['team1'] = result_match.group(1).strip()
            info['team1_score'] = int(result_match.group(2))
            info['team1_wickets'] = int(result_match.group(3))
            info['team1_overs'] = float(result_match.group(4))
            info['team2'] = result_match.group(5).strip()
            info['winning_team'] = result_match.group(5).strip()
            info['result_margin'] = int(result_match.group(6))
            info['result_type'] = result_match.group(7).strip()
        
        # Tournament
        tournament_match = re.search(r'Tournament\s+([^\n]+)', text)
        if tournament_match:
            info['tournament'] = tournament_match.group(1).strip()
        
        # Date and time
        date_match = re.search(r'Date&Time\s+([^\n]+)', text)
        if date_match:
            info['date_time'] = date_match.group(1).strip()
        
        # Venue
        venue_match = re.search(r'Venue\s+([^\n]+)', text)
        if venue_match:
            info['venue'] = venue_match.group(1).strip()
        
        # Toss
        toss_match = re.search(r'Toss\s+(\w+.*?)\s+Opted\s+To\s+(Bat|Bowl)', text, re.IGNORECASE)
        if toss_match:
            info['toss_winner'] = toss_match.group(1).strip()
            info['toss_choice'] = toss_match.group(2).strip()
        
        return info
    
    def parse_dismissal(self, dismissal_text: str) -> Tuple[str, Optional[str], Optional[str]]:
        """
        Parse dismissal notation to extract type and related players
        Returns: (dismissal_type, bowler_name, fielder_name)
        """
        dismissal_text = dismissal_text.strip()
        
        # not out
        if 'notout' in dismissal_text.lower() or 'not out' in dismissal_text.lower():
            return ('not_out', None, None)
        
        # bowled: b BowlerName
        bowled_match = re.search(r'\bb\s+(\w+(?:\s+\w+)*?)(?:\s|$)', dismissal_text, re.IGNORECASE)
        if bowled_match:
            bowler = bowled_match.group(1).strip()
            return ('bowled', bowler, None)
        
        # caught and bowled: c&b BowlerName
        candb_match = re.search(r'c&b\s+(\w+(?:\s+\w+)*?)(?:\s|$)', dismissal_text, re.IGNORECASE)
        if candb_match:
            bowler = candb_match.group(1).strip()
            return ('caught', bowler, bowler)  # fielder is also the bowler
        
        # caught: c FielderName b BowlerName
        caught_match = re.search(r'c\s+(\w+(?:\s+\w+)*?)\s+b\s+(\w+(?:\s+\w+)*?)(?:\s|$)', dismissal_text, re.IGNORECASE)
        if caught_match:
            fielder = caught_match.group(1).strip()
            bowler = caught_match.group(2).strip()
            return ('caught', bowler, fielder)
        
        # lbw: lbw b BowlerName
        lbw_match = re.search(r'lbw\s+b\s+(\w+(?:\s+\w+)*?)(?:\s|$)', dismissal_text, re.IGNORECASE)
        if lbw_match:
            bowler = lbw_match.group(1).strip()
            return ('lbw', bowler, None)
        
        # run out: run out (FielderName) or just "run out"
        runout_match = re.search(r'run\s+out(?:\s*\(([^)]+)\))?', dismissal_text, re.IGNORECASE)
        if runout_match:
            fielder = runout_match.group(1).strip() if runout_match.group(1) else None
            return ('runout', None, fielder)
        
        # stumped: st FielderName b BowlerName
        stumped_match = re.search(r'st\s+(\w+(?:\s+\w+)*?)\s+b\s+(\w+(?:\s+\w+)*?)(?:\s|$)', dismissal_text, re.IGNORECASE)
        if stumped_match:
            fielder = stumped_match.group(1).strip()
            bowler = stumped_match.group(2).strip()
            return ('stumped', bowler, fielder)
        
        # mankad: mankad by BowlerName or just "mankad"
        mankad_match = re.search(r'mankad(?:\s+by\s+(\w+(?:\s+\w+)*?))?(?:\s|$)', dismissal_text, re.IGNORECASE)
        if mankad_match:
            bowler = mankad_match.group(1).strip() if mankad_match.group(1) else None
            return ('mankad', bowler, None)
        
        # Default: other
        return ('other', None, None)
    
    def parse_batting_scorecard(self, text: str, innings: int = 1) -> List[Dict[str, Any]]:
        """
        Extract batting statistics from scorecard
        innings: 1 for 1st innings, 2 for 2nd innings
        """
        stats = []
        
        # Find the innings section
        if innings == 1:
            pattern = r'1stInningsScorecard.*?(?=2ndInningsScorecard|OverComparison|---PAGE_BREAK---|$)'
        else:
            pattern = r'2ndInningsScorecard.*?(?=OverComparison|---PAGE_BREAK---|$)'
        
        innings_text = re.search(pattern, text, re.DOTALL | re.IGNORECASE)
        if not innings_text:
            return stats
        
        innings_section = innings_text.group(0)
        
        # Split into batting and bowling sections
        # Batting table comes first, bowling after "Bowler"
        lines = innings_section.split('\n')
        
        in_batting_table = False
        for i, line in enumerate(lines):
            line = line.strip()
            
            # Start of batting table
            if re.match(r'^\w+.*?\s+R\s+B\s+4s\s+6s\s+SR', line, re.IGNORECASE):
                in_batting_table = True
                continue
            
            # End of batting table (when we see "Extras" or "Bowler")
            if re.match(r'(Extras|Bowler|Total|FallOf)', line, re.IGNORECASE):
                in_batting_table = False
                continue
            
            # Parse batting row
            if in_batting_table and line:
                # Pattern: PlayerName dismissal_info runs balls 4s 6s sr
                # e.g., "BinayaKhadka bRodin 9 8 1 0 112.5"
                match = re.match(r'^([A-Z][a-zA-Z]+(?:\s+[A-Z][a-zA-Z]+)*?)\s+([a-z].*?)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+\.?\d*)', line)
                if match:
                    player_name = match.group(1).strip()
                    dismissal_text = match.group(2).strip()
                    runs = int(match.group(3))
                    balls = int(match.group(4))
                    fours = int(match.group(5))
                    sixes = int(match.group(6))
                    sr = float(match.group(7))
                    
                    dismissal_type, bowler, fielder = self.parse_dismissal(dismissal_text)
                    
                    stat = {
                        'player_name': player_name,
                        'runs': runs,
                        'balls': balls,
                        'fours': fours,
                        'sixes': sixes,
                        'strike_rate': sr,
                        'dismissal_type': dismissal_type,
                        'bowler_name': bowler,
                        'fielder_name': fielder,
                        'innings': innings
                    }
                    stats.append(stat)
        
        return stats
    
    def parse_bowling_scorecard(self, text: str, innings: int = 1) -> List[Dict[str, Any]]:
        """Extract bowling statistics from scorecard"""
        stats = []
        
        # Find the innings section
        if innings == 1:
            pattern = r'1stInningsScorecard.*?(?=2ndInningsScorecard|OverComparison|---PAGE_BREAK---|$)'
        else:
            pattern = r'2ndInningsScorecard.*?(?=OverComparison|---PAGE_BREAK---|$)'
        
        innings_text = re.search(pattern, text, re.DOTALL | re.IGNORECASE)
        if not innings_text:
            return stats
        
        innings_section = innings_text.group(0)
        lines = innings_section.split('\n')
        
        in_bowling_table = False
        for line in lines:
            line = line.strip()
            
            # Start of bowling table
            if re.match(r'Bowler\s+O\s+M\s+R\s+W\s+Eco', line, re.IGNORECASE):
                in_bowling_table = True
                continue
            
            # End of bowling table
            if re.match(r'(Total|Match|Download)', line, re.IGNORECASE):
                in_bowling_table = False
                continue
            
            # Parse bowling row
            if in_bowling_table and line:
                # Pattern: BowlerName overs maidens runs wickets economy ...
                # e.g., "AmitChaulagain 2.0 0 18 0 9.0 4 1 1 1 1"
                match = re.match(r'^([A-Z][a-zA-Z]+(?:\s+[A-Z][a-zA-Z]+)*?)\s+([\d.]+)\s+(\d+)\s+(\d+)\s+(\d+)\s+([\d.]+)', line)
                if match:
                    bowler_name = match.group(1).strip()
                    overs = float(match.group(2))
                    maidens = int(match.group(3))
                    runs = int(match.group(4))
                    wickets = int(match.group(5))
                    economy = float(match.group(6))
                    
                    stat = {
                        'bowler_name': bowler_name,
                        'overs': overs,
                        'maidens': maidens,
                        'runs': runs,
                        'wickets': wickets,
                        'economy': economy,
                        'innings': innings
                    }
                    stats.append(stat)
        
        return stats
    
    def parse(self) -> Dict[str, Any]:
        """Parse entire PDF and return structured data"""
        text = self.extract_text()
        
        result = {
            'match_info': self.parse_match_info(text),
            'innings_1_batting': self.parse_batting_scorecard(text, 1),
            'innings_1_bowling': self.parse_bowling_scorecard(text, 1),
            'innings_2_batting': self.parse_batting_scorecard(text, 2),
            'innings_2_bowling': self.parse_bowling_scorecard(text, 2),
        }
        
        return result


# Test the parser
if __name__ == '__main__':
    parser = STUMPSPDFParser('/Users/sarikgautam/Documents/GNCPL/LovDev/goldcoast-cricket-feud/hhvsys.pdf')
    result = parser.parse()
    
    import json
    print(json.dumps(result, indent=2))

File: src/lib/pdfParser.ts
import * as pdfjsLib from 'pdfjs-dist';

// Configure worker for Vite
pdfjsLib.GlobalWorkerOptions.workerSrc = new URL(
  'pdfjs-dist/build/pdf.worker.min.mjs',
  import.meta.url
).toString();

export interface DismissalInfo {
  type: 'not_out' | 'bowled' | 'caught' | 'lbw' | 'runout' | 'stumped' | 'mankad' | 'retired_hurt' | 'other';
  bowler?: string;
  fielder?: string;
}

export interface BattingStats {
  playerName: string;
  runs: number;
  balls: number;
  fours: number;
  sixes: number;
  strikeRate: number;
  dismissal: DismissalInfo;
  innings: number;
}

export interface BowlingStats {
  bowlerName: string;
  overs: number;
  maidens: number;
  runs: number;
  wickets: number;
  economy: number;
  innings: number;
}

export interface MatchInfo {
  team1?: string;
  team2?: string;
  team1Score?: number;
  team1Wickets?: number;
  team1Overs?: number;
  winningTeam?: string;
  resultMargin?: number;
  resultType?: string;
  tournament?: string;
  dateTime?: string;
  venue?: string;
  tossWinner?: string;
  tossChoice?: string;
}

export interface ParsedScorecard {
  matchInfo: MatchInfo;
  innings1Batting: BattingStats[];
  innings1Bowling: BowlingStats[];
  innings2Batting: BattingStats[];
  innings2Bowling: BowlingStats[];
}

/**
 * Parse dismissal notation to extract type and related players
 */
function parseDismissal(dismissalText: string): DismissalInfo {
  const text = dismissalText.trim().toLowerCase();

  // not out
  if (text.includes('notout') || text.includes('not out')) {
    return { type: 'not_out' };
  }

  // caught and bowled: c&b BowlerName
  const candbMatch = dismissalText.match(/c&b\s*([A-Za-z]+(?:\s+[A-Za-z]+)*)/i);
  if (candbMatch) {
    const bowler = candbMatch[1].trim();
    return { type: 'caught', bowler, fielder: bowler };
  }

  // caught: c FielderName b BowlerName (with spaces)
  const caughtMatchSpaced = dismissalText.match(/c\s+([A-Za-z]+(?:\s+[A-Za-z]+)*?)\s+b\s+([A-Za-z]+(?:\s+[A-Za-z]+)*)/i);
  if (caughtMatchSpaced) {
    const fielder = caughtMatchSpaced[1].trim();
    const bowler = caughtMatchSpaced[2].trim();
    return { type: 'caught', bowler, fielder };
  }

  // caught: cFielderNamebBowlerName (without spaces - from PDF extraction)
  const caughtMatchNoSpace = dismissalText.match(/^c([A-Za-z]+)b([A-Za-z]+)$/i);
  if (caughtMatchNoSpace) {
    const fielder = caughtMatchNoSpace[1];
    const bowler = caughtMatchNoSpace[2];
    return { type: 'caught', bowler, fielder };
  }

  // lbw: lbw b BowlerName or just lbw BowlerName or lbwBowlerName
  const lbwMatch = dismissalText.match(/lbw\s*(?:b\s*)?([A-Za-z]+(?:\s+[A-Za-z]+)*)/i);
  if (lbwMatch) {
    const bowler = lbwMatch[1].trim();
    return { type: 'lbw', bowler };
  }

  // bowled: b BowlerName (with space or without)
  const bowledMatch = dismissalText.match(/\bb\s*([A-Za-z]+(?:\s+[A-Za-z]+)*)/i);
  if (bowledMatch) {
    const bowler = bowledMatch[1].trim();
    return { type: 'bowled', bowler };
  }

  // run out: run out (FielderName) or runout(FielderName)
  const runoutMatch = text.match(/run\s*out(?:\s*\(([^)]+)\))?/i);
  if (runoutMatch) {
    const fielder = runoutMatch[1]?.trim();
    return { type: 'runout', fielder };
  }

  // stumped: st FielderName b BowlerName (with spaces)
  const stumpedMatchSpaced = dismissalText.match(/st\s+([A-Za-z]+(?:\s+[A-Za-z]+)*?)\s+b\s+([A-Za-z]+(?:\s+[A-Za-z]+)*)/i);
  if (stumpedMatchSpaced) {
    const fielder = stumpedMatchSpaced[1].trim();
    const bowler = stumpedMatchSpaced[2].trim();
    return { type: 'stumped', bowler, fielder };
  }

  // stumped: stFielderNamebBowlerName (without spaces)
  const stumpedMatchNoSpace = dismissalText.match(/^st([A-Za-z]+)b([A-Za-z]+)$/i);
  if (stumpedMatchNoSpace) {
    const fielder = stumpedMatchNoSpace[1];
    const bowler = stumpedMatchNoSpace[2];
    return { type: 'stumped', bowler, fielder };
  }

  // mankad: mankad by BowlerName or just "mankad"
  const mankadMatch = dismissalText.match(/mankad(?:\s+by\s+([A-Za-z]+(?:\s+[A-Za-z]+)*))?/i);
  if (mankadMatch) {
    const bowler = mankadMatch[1]?.trim();
    return { type: 'mankad', bowler };
  }

  // retired hurt
  if (text.includes('retired') && text.includes('hurt')) {
    return { type: 'retired_hurt' };
  }

  // Default: other
  console.log(`Could not parse dismissal: "${dismissalText}"`);
  return { type: 'other' };
}

/**
 * Extract match information from page 1
 */
function parseMatchInfo(text: string): MatchInfo {
  const info: MatchInfo = {};

  // Match result
  const resultMatch = text.match(/(\w+.*?)\s+(\d+)-(\d+)in([\d.]+)overs\s+(\w+.*?)\s+won\s+by\s+(\d+)\s+(wickets|runs)/i);
  if (resultMatch) {
    info.team1 = resultMatch[1].trim();
    info.team1Score = parseInt(resultMatch[2]);
    info.team1Wickets = parseInt(resultMatch[3]);
    info.team1Overs = parseFloat(resultMatch[4]);
    info.team2 = resultMatch[5].trim();
    info.winningTeam = resultMatch[5].trim();
    info.resultMargin = parseInt(resultMatch[6]);
    info.resultType = resultMatch[7].trim();
  }

  // Tournament
  const tournamentMatch = text.match(/Tournament\s+([^\n]+)/i);
  if (tournamentMatch) {
    info.tournament = tournamentMatch[1].trim();
  }

  // Date and time
  const dateMatch = text.match(/Date&Time\s+([^\n]+)/i);
  if (dateMatch) {
    info.dateTime = dateMatch[1].trim();
  }

  // Venue
  const venueMatch = text.match(/Venue\s+([^\n]+)/i);
  if (venueMatch) {
    info.venue = venueMatch[1].trim();
  }

  // Toss
  const tossMatch = text.match(/Toss\s+(\w+.*?)\s+Opted\s+To\s+(Bat|Bowl)/i);
  if (tossMatch) {
    info.tossWinner = tossMatch[1].trim();
    info.tossChoice = tossMatch[2].trim();
  }

  return info;
}

/**
 * Parse batting scorecard for a specific innings
 */
function parseBattingScorecard(text: string, innings: number): BattingStats[] {
  const stats: BattingStats[] = [];

  // The PDF might not preserve exact section headers, so let's search more flexibly
  // Look for the team name patterns and batting stats that follow
  
  // First, try to find sections by looking for the exact headers
  let pattern = innings === 1
    ? /1st\s*Innings\s*Scorecard([\s\S]*?)(?=2nd\s*Innings\s*Scorecard|Over\s*Comparison|$)/i
    : /2nd\s*Innings\s*Scorecard([\s\S]*?)(?=Over\s*Comparison|$)/i;

  let inningsMatch = text.match(pattern);
  
  if (!inningsMatch) {
    console.log(`Trying alternative pattern for innings ${innings}`);
    // Alternative: look for team name followed by R B 4s 6s SR pattern
    // Split by "---PAGE_BREAK---" and look at individual pages
    const pages = text.split('---PAGE_BREAK---');
    console.log(`Found ${pages.length} pages`);
    
    // Page 2 should have 1st innings (index 1), Page 3 should have 2nd innings (index 2)
    const pageIndex = innings === 1 ? 1 : 2;
    
    if (pages.length > pageIndex) {
      console.log(`Using page ${pageIndex + 1} for innings ${innings}`);
      console.log(`Page ${pageIndex + 1} preview:`, pages[pageIndex].substring(0, 500));
      
      // Look for the R B 4s 6s SR pattern on this page
      const rbPattern = /R\s+B\s+4s\s+6s\s+SR/i;
      if (rbPattern.test(pages[pageIndex])) {
        // Parse this page as the innings section
        return parseBattingFromText(pages[pageIndex], innings);
      }
    }
    
    console.log(`No innings ${innings} section found in text`);
    return stats;
  }

  return parseBattingFromText(inningsMatch[1], innings);
}

function parseBattingFromText(inningsSection: string, innings: number): BattingStats[] {
  const stats: BattingStats[] = [];
  
  console.log(`Innings ${innings} section length:`, inningsSection.length);
  console.log(`Innings ${innings} preview:`, inningsSection.substring(0, 300));
  
  // Split text into words/tokens to parse more flexibly
  const tokens = inningsSection.split(/\s+/).filter(t => t.length > 0);
  
  // Find where batting stats start (after R B 4s 6s SR header)
  let startIndex = -1;
  for (let i = 0; i < tokens.length - 4; i++) {
    if (tokens[i] === 'R' && tokens[i+1] === 'B' && tokens[i+2] === '4s' && tokens[i+3] === '6s' && tokens[i+4] === 'SR') {
      startIndex = i + 5;
      break;
    }
  }
  
  if (startIndex === -1) {
    console.log('Could not find batting stats header (R B 4s 6s SR)');
    return stats;
  }
  
  console.log('Found batting header at token index:', startIndex);
  
  // Parse batting rows - each row has: PlayerName dismissal runs balls 4s 6s SR
  // Look for pattern: Name (capitalized) followed by dismissal text, then 5 numbers
  for (let i = startIndex; i < tokens.length - 6; i++) {
    const token = tokens[i];
    
    // Stop at section markers
    if (/^(Extras|Bowler|Total|FallOf)/i.test(token)) {
      break;
    }
    
    // Check if this looks like a player name (starts with capital letter)
    if (/^[A-Z][a-z]+/.test(token)) {
      // Collect potential player name (could be multi-word like "Ramesh Subedi")
      let playerName = token;
      let nextIdx = i + 1;
      
      // Check if next token is also part of name (capitalized)
      while (nextIdx < tokens.length && /^[A-Z][a-z]+/.test(tokens[nextIdx])) {
        playerName += ' ' + tokens[nextIdx];
        nextIdx++;
      }
      
      // Next should be dismissal info (starts with lowercase)
      if (nextIdx >= tokens.length || !/^[a-z&]/.test(tokens[nextIdx])) {
        continue;
      }
      
      let dismissalText = tokens[nextIdx];
      nextIdx++;
      
      // Dismissal might be multi-token (e.g., "c", "Fielder", "b", "Bowler")
      // or single token (e.g., "bRodin", "notout")
      // Collect until we hit a number
      while (nextIdx < tokens.length && !/^\d+$/.test(tokens[nextIdx])) {
        dismissalText += tokens[nextIdx];
        nextIdx++;
      }
      
      // Now we should have 5 numbers: runs, balls, 4s, 6s, SR
      if (nextIdx + 4 >= tokens.length) {
        continue;
      }
      
      const runsStr = tokens[nextIdx];
      const ballsStr = tokens[nextIdx + 1];
      const foursStr = tokens[nextIdx + 2];
      const sixesStr = tokens[nextIdx + 3];
      const srStr = tokens[nextIdx + 4];
      
      // Validate they're all numbers
      if (!/^\d+$/.test(runsStr) || !/^\d+$/.test(ballsStr) || 
          !/^\d+$/.test(foursStr) || !/^\d+$/.test(sixesStr) || 
          !/^[\d.]+$/.test(srStr)) {
        continue;
      }
      
      const runs = parseInt(runsStr);
      const balls = parseInt(ballsStr);
      const fours = parseInt(foursStr);
      const sixes = parseInt(sixesStr);
      const strikeRate = parseFloat(srStr);
      
      const dismissal = parseDismissal(dismissalText);
      
      stats.push({
        playerName,
        runs,
        balls,
        fours,
        sixes,
        strikeRate,
        dismissal,
        innings,
      });
      
      console.log(`Parsed batting stat for ${playerName}: ${runs}(${balls}) - ${dismissalText}`);
      
      // Move index forward past this entry
      i = nextIdx + 4;
    }
  }

  console.log(`Found ${stats.length} batting stats for innings ${innings}`);
  return stats;
}

/**
 * Parse bowling scorecard for a specific innings
 */
function parseBowlingScorecard(text: string, innings: number): BowlingStats[] {
  const stats: BowlingStats[] = [];

  // Try exact pattern first
  let pattern = innings === 1
    ? /1st\s*Innings\s*Scorecard([\s\S]*?)(?=2nd\s*Innings\s*Scorecard|Over\s*Comparison|$)/i
    : /2nd\s*Innings\s*Scorecard([\s\S]*?)(?=Over\s*Comparison|$)/i;

  let inningsMatch = text.match(pattern);
  
  if (!inningsMatch) {
    // Alternative: use page-based extraction
    const pages = text.split('---PAGE_BREAK---');
    const pageIndex = innings === 1 ? 1 : 2;
    
    if (pages.length > pageIndex) {
      return parseBowlingFromText(pages[pageIndex], innings);
    }
    
    console.log(`No innings ${innings} section found for bowling`);
    return stats;
  }

  return parseBowlingFromText(inningsMatch[1], innings);
}

function parseBowlingFromText(inningsSection: string, innings: number): BowlingStats[] {
  const stats: BowlingStats[] = [];
  const tokens = inningsSection.split(/\s+/).filter(t => t.length > 0);

  // Find bowling header: Bowler O M R W Eco
  let startIndex = -1;
  for (let i = 0; i < tokens.length - 5; i++) {
    if (tokens[i] === 'Bowler' && tokens[i+1] === 'O' && tokens[i+2] === 'M' && 
        tokens[i+3] === 'R' && tokens[i+4] === 'W' && tokens[i+5] === 'Eco') {
      startIndex = i + 6;
      break;
    }
  }

  if (startIndex === -1) {
    console.log('Could not find bowling header');
    return stats;
  }

  console.log('Found bowling header at token index:', startIndex);

  // Parse bowling rows: BowlerName overs maidens runs wickets economy (plus extra columns)
  for (let i = startIndex; i < tokens.length - 5; i++) {
    const token = tokens[i];
    
    // Stop at section markers
    if (/^(Total|Match|Download|FallOf|Over)/i.test(token)) {
      break;
    }
    
    // Check for bowler name (capitalized)
    if (/^[A-Z][a-z]+/.test(token)) {
      let bowlerName = token;
      let nextIdx = i + 1;
      
      // Multi-word name
      while (nextIdx < tokens.length && /^[A-Z][a-z]+/.test(tokens[nextIdx])) {
        bowlerName += ' ' + tokens[nextIdx];
        nextIdx++;
      }
      
      // Next should be numbers: overs maidens runs wickets economy
      if (nextIdx + 4 >= tokens.length) {
        continue;
      }
      
      const oversStr = tokens[nextIdx];
      const maidensStr = tokens[nextIdx + 1];
      const runsStr = tokens[nextIdx + 2];
      const wicketsStr = tokens[nextIdx + 3];
      const economyStr = tokens[nextIdx + 4];
      
      // Validate
      if (!/^[\d.]+$/.test(oversStr) || !/^\d+$/.test(maidensStr) || 
          !/^\d+$/.test(runsStr) || !/^\d+$/.test(wicketsStr) || 
          !/^[\d.]+$/.test(economyStr)) {
        continue;
      }
      
      const overs = parseFloat(oversStr);
      const maidens = parseInt(maidensStr);
      const runs = parseInt(runsStr);
      const wickets = parseInt(wicketsStr);
      const economy = parseFloat(economyStr);

      stats.push({
        bowlerName,
        overs,
        maidens,
        runs,
        wickets,
        economy,
        innings,
      });
      
      console.log(`Parsed bowling stat for ${bowlerName}: ${overs} overs, ${runs}R, ${wickets}W`);
      
      i = nextIdx + 4;
    }
  }

  console.log(`Found ${stats.length} bowling stats for innings ${innings}`);
  return stats;
}

/**
 * Extract text from PDF file
 */
async function extractPDFText(file: File): Promise<string> {
  try {
    console.log('Starting PDF extraction for file:', file.name, 'size:', file.size);
    
    if (!file.size) {
      throw new Error('PDF file is empty');
    }

    const arrayBuffer = await file.arrayBuffer();
    console.log('Converted file to ArrayBuffer, size:', arrayBuffer.byteLength);

    const pdf = await pdfjsLib.getDocument({ data: arrayBuffer }).promise;
    console.log('PDF loaded successfully, pages:', pdf.numPages);

    if (pdf.numPages === 0) {
      throw new Error('PDF has no pages');
    }

    let fullText = '';

    for (let i = 1; i <= pdf.numPages; i++) {
      const page = await pdf.getPage(i);
      const textContent = await page.getTextContent();
      const pageText = textContent.items
        .map((item: any) => item.str)
        .join(' ');
      fullText += pageText + '\n---PAGE_BREAK---\n';
      console.log(`Extracted page ${i}, text length: ${pageText.length}`);
    }

    console.log('PDF text extraction complete, total length:', fullText.length);
    
    if (!fullText.trim()) {
      throw new Error('No text could be extracted from PDF - file might be scanned image');
    }

    return fullText;
  } catch (error) {
    console.error('PDF extraction error:', error);
    throw new Error(`Failed to extract text from PDF: ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Parse STUMPS cricket scorecard PDF
 */
export async function parsePDFScorecard(file: File): Promise<ParsedScorecard> {
  console.log('Starting PDF parse for:', file.name);
  const text = await extractPDFText(file);
  
  console.log('Extracted text length:', text.length);
  console.log('Text preview:', text.substring(0, 500));

  const matchInfo = parseMatchInfo(text);
  console.log('Match info:', matchInfo);

  const innings1Batting = parseBattingScorecard(text, 1);
  const innings1Bowling = parseBowlingScorecard(text, 1);
  const innings2Batting = parseBattingScorecard(text, 2);
  const innings2Bowling = parseBowlingScorecard(text, 2);

  console.log('Parse complete - Batting stats:', {
    innings1: innings1Batting.length,
    innings2: innings2Batting.length,
  });

  return {
    matchInfo,
    innings1Batting,
    innings1Bowling,
    innings2Batting,
    innings2Bowling,
  };
}
